Spec: Add implementation note on `current-snapshot-id` (#12334)
apache/iceberg: format/spec.md

| **`engine-name`**        | "spark"    | Name of the engine that created the snapshot                    |
| **`engine-version`**     | "3.5.4"    | Version of the engine that created the snapshot                 |

### Encoding of `current-snapshot-id`

> For the Java implementation, `-1` has meant "no current snapshot" in the past and is equivalent to missing/null. However, this has never been formalized in the spec. When reading this field, accept `-1` as `null`. For Java, ≥V3 metadata will write `null` instead of `-1` in the case there is no current snapshot.

Reviewer: ```suggestion
The Java implementation writes `-1` for  "no current snapshot" with V1 and V2 tables and considers this equivalent to missing/null. This has never been formalized in the spec but for compatibility other implementations can accept `-1` as `null`. The Java implementation will no longer write `-1` and will use null for "no current snapshot for all tables with a version greater than or equal to V3```

Just reworded things a bit but feel free to ignore.

Author: I like it. I copied some sentences exactly from the mail to capture it as good as possible, but I like the rewording.

> The Java implementation writes `-1` for  "no current snapshot" with V1 and V2 tables and considers this equivalent to missing/null. This has never been formalized in the spec but for compatibility other implementations can accept `-1` as `null`. The Java implementation will no longer write `-1` and will use null for "no current snapshot for all tables with a version greater than or equal to V3

Reviewer 2: Puncutation nits: 

- extra space between `for "no current snapshot"`
- comma before 'but' (second is complete sentence)
- period at end
- should we be consistent whether `null` is code-denoted or not

Author: ```suggestion
The Java implementation writes `-1` for "no current snapshot" with V1 and V2 tables and considers this equivalent to omitted or `null`. This has never been formalized in the spec but for compatibility, other implementations can accept `-1` as `null`. The Java implementation will no longer write `-1` and will use `null` for "no current snapshot" for all tables with a version greater than or equal to V3.
```

Reviewer 2: Looks great, thanks !

| **`engine-name`**        | "spark"    | Name of the engine that created the snapshot                    |
| **`engine-version`**     | "3.5.4"    | Version of the engine that created the snapshot                 |

### Assignment of Snapshot IDs and `current-snapshot-id`

> Writers should produce positive values for snapshot ids in a manner that minimizes the probability of id collisions and should verify the id does not conflict with existing snapshots. Producing snapshot ids based on timestamps alone is not recommended as it increases the potential for collisions.

Reviewer 3: is generating a positive value a requirement here?  It seems that it might be more of than an implementation note that "-1" should be reserved because it has special meaning prior to V3?

Reviewer 4: There's nothing in the spec that that explicitly prohibits negative values, but the approach by the reference implementation intentionally avoids negative snapshot ids. There is commentary on this in #57 and #2506, so I think there is an expectation even if not explicitly called out.

Reviewer 3: Right, so I think -1 can't just be an implementation detail with a specific version modifier saying it is ok to produce. if we don't think negative numbers break anything then I think it seems fine to leave as an implementation detail.

Reviewer 5: I don't remember it ever being `System.currentTimeMillis`. Was it really?

Reviewer 4: Yes, the original was just current millis, but it had issues (particularly with tests having collisions).

> Writers should produce positive values for snapshot ids in a manner that minimizes the probability of id collisions and should verify the id does not conflict with existing snapshots.  Producing snapshot ids based on timestamps alone is not recommended as it increases the potential for collisions.  

Author: ```suggestion
Writers should produce positive values for snapshot ids in a manner that minimizes the probability of id collisions and should verify the id does not conflict with existing snapshots. Producing snapshot ids based on timestamps alone is not recommended as it increases the potential for collisions.
```

The reference implementation uses a type 4 uuid and XORs the 4 most significant bytes with the 4 least significant bytes then ANDs with the maximum long value to arrive at a pseudo-random snapshot id with a low probability of collision.

> The reference Java implementation writes `-1` for "no current snapshot" with V1 and V2 tables and considers this equivalent to omitted or `null`. This has never been formalized in the spec but for compatibility, other implementations can accept `-1` as `null`. Java will no longer write `-1` and will use `null` for "no current snapshot" for all tables with a version greater than or equal to V3.

Author: ```suggestion
Java writes `-1` for "no current snapshot" with V1 and V2 tables and considers this equivalent to omitted or `null`. This has never been formalized in the spec, but for compatibility, other implementations can accept `-1` as `null`. Java will no longer write `-1` and will use `null` for "no current snapshot" for all tables with a version greater than or equal to V3.
```

> The reference Java implementation writes `-1` for "no current snapshot" with V1 and V2 tables and considers this equivalent to omitted or `null`. This has never been formalized in the spec but for compatibility, other implementations can accept `-1` as `null`. Java will no longer write `-1` and will use `null` for "no current snapshot" for all tables with a version greater than or equal to V3.

Author: @szehon-ho Good point; thanks for iterating on this. What do you think of this form? It doesn't read well to call out `The reference Java implementation` twice. WDYT?

> The reference Java implementation writes `-1` for "no current snapshot" with V1 and V2 tables and considers this equivalent to omitted or `null`. This has never been formalized in the spec but for compatibility, other implementations can accept `-1` as `null`. Java will no longer write `-1` and will use `null` for "no current snapshot" for all tables with a version greater than or equal to V3.

Reviewer 2: Should we just put 'The reference Java implementation' in the first pargraph then the first time we mention it?

`The reference Java implementation uses a type 4 uuid and XORs the 4 most significant bytes with the 4 least significant bytes then ANDs with the maximum long value to arrive at a pseudo-random snapshot id with a low probability of collision.`

And then can use Java threafter?

> The reference Java implementation writes `-1` for "no current snapshot" with V1 and V2 tables and considers this equivalent to omitted or `null`. This has never been formalized in the spec but for compatibility, other implementations can accept `-1` as `null`. Java will no longer write `-1` and will use `null` for "no current snapshot" for all tables with a version greater than or equal to V3.

Author: Yes, I think this aligns with the current state: https://github.com/apache/iceberg/pull/12334/files#diff-36347a47c3bf67ea2ef6309ea96201814032d21bb5f162dfae4045508c15588aR1761-R1763

> The Java implementation writes `-1` for "no current snapshot" with V1 and V2 tables and considers this equivalent to omitted or `null`. This has never been formalized in the spec but for compatibility, other implementations can accept `-1` as `null`. The Java implementation will no longer write `-1` and will use `null` for "no current snapshot" for all tables with a version greater than or equal to V3.

Reviewer 6: It is a little bit confusing to see `The reference implementation` and `The Java implementation` at the same time. What about consolidate them with `The Java reference implementation`?

Author: What do you think of the following:
```suggestion
The reference Java implementation writes `-1` for "no current snapshot" with V1 and V2 tables and considers this equivalent to omitted or `null`. This has never been formalized in the spec but for compatibility, other implementations can accept `-1` as `null`. Java will no longer write `-1` and will use `null` for "no current snapshot" for all tables with a version greater than or equal to V3.
```
